Repo URL:  https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>/blob/main/slime_sample_notebook.ipynb
Commit ID: 6812f559ba1f8b7054064806d6866a57976b3ee6

In [1]:
# Imports section
import pandas as pd 
import numpy as np 
from sklearn.model_selection import train_test_split, cross_val_score  
from sklearn.linear_model import LinearRegression 
from sklearn.preprocessing import PolynomialFeatures 
from sklearn.metrics import mean_squared_error


## Part 1. Loading the dataset

In [3]:
# Using pandas load the dataset (load remotely, not locally)
url = 'https://raw.githubusercontent.com/COMPSCI-383-Fall2024/assignment-4-equation-of-a-slime-<redacted>/refs/heads/main/science_data_large.csv?token=GHSAT0AAAAAACYTEJ3MQQVOUYCWK552DET4ZYXWV4Q'
data = pd.read_csv(url)
# Output the first 15 rows of the data
print(data.head(15))
# Display a summary of the table information (number of datapoints, etc.)
data.info()

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [4]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = data[['Mols KCL', 'Temperature °C']]  
y = data['Size nm^3'] 
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

print(f"Training set shape: X_train: {X_train.shape}, y_train: {y_train.shape}")
print(f"Test set shape: X_test: {X_test.shape}, y_test: {y_test.shape}")

Training set shape: X_train: (900, 2), y_train: (900,)
Test set shape: X_test: (100, 2), y_test: (100,)


## Part 3. Perform a Linear Regression

In [5]:
# Use sklearn to train a model on the training set

linear_model = LinearRegression()
linear_model.fit(X_train, y_train)

print("Coefficients:", linear_model.coef_)
print("Intercept:", linear_model.intercept_)

# Create a sample datapoint and predict the output of that sample with the trained model

sample_data = [[2, 25]]
predicted_size = linear_model.predict(sample_data)

print(f"Predicted size of slime for 2 mols of KCl and 25°C: {predicted_size[0]}")

# Report on the score for that model, in your own words (markdown, not code) explain what the score means

score = linear_model.score(X_test, y_test)

print(f"Model R^2 Score: {score}")

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX



Coefficients: [1032.69506649  866.14641337]
Intercept: -409391.4795834082
Predicted size of slime for 2 mols of KCl and 25°C: -385672.4291161386
Model R^2 Score: 0.8552472077276096


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


### Model Equation

The equation for our model is: 

$$
h(x) = \text{intercept} + \text{coef}_1 \cdot \text{KCl} + \text{coef}_2 \cdot \text{Temperature}
$$

Where:
- $\text{intercept}$ = [-409391.4795834082]
- $\text{coef}_1$ = [1032.69506649]
- $\text{coef}_2$ = [866.14641337]



Sample equation: $E = mc^2$

## Part 4. Use Cross Validation

In [ ]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
cv_scores = cross_val_score(linear_model, X_train, y_train, cv=5)

print("Cross-validation scores for each fold:", cv_scores)
print("Mean cross-validation score:", cv_scores.mean())
print("Standard deviation of cross-validation scores:", cv_scores.std())
# Report on their finding and their significance


Cross-validation scores for each fold: [0.86226163 0.81982226 0.88938198 0.86663176 0.85729958]
Mean cross-validation score: 0.8590794402339516
Standard deviation of cross-validation scores: 0.022493119105236842


### Cross-Validation Findings:

The model achieved a mean cross-validation score of `0.8590794402339516`, which suggests that the linear regression model performs reasonably well in predicting the size of the slime based on the KCl concentration and temperature. The standard deviation of `0.022493119105236842` indicates that the model's performance is fairly consistent across different folds, showing minimal variability.

A lower standard deviation signifies that the model's performance is stable, meaning it is likely to generalize well to unseen data. If the standard deviation were higher, it would suggest that the model's predictions may vary more depending on the specific split of the data, potentially indicating overfitting or underfitting.

Overall, the cross-validation results suggest that the model is reliable for predicting the growth of slime under different conditions.


## Part 5. Using Polynomial Regression

In [7]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2

poly = PolynomialFeatures(degree=2)

X_poly_train = poly.fit_transform(X_train)
X_poly_test = poly.transform(X_test)

poly_model = LinearRegression()
poly_model.fit(X_poly_train, y_train)

y_poly_pred = poly_model.predict(X_poly_test)

mse = mean_squared_error(y_test, y_poly_pred)
r2 = poly_model.score(X_poly_test, y_test)

print(f"Mean Squared Error: {mse}")
print(f"R^2 Score: {r2}")

# Report on the metrics and output the resultant equation as you did in Part 3.

Mean Squared Error: 1.4693515832582168e-08
R^2 Score: 1.0


### Polynomial Regression Findings:

The polynomial regression model (degree 2) achieved a **Mean Squared Error** of `1.4693515832582168e-08` and an **R² Score** of `1.0`. Compared to the linear regression model, this suggests that the polynomial regression model fits the data perfectly, explaining all of the variance in the target variable (slime size).

However, an **R² score** of `1.0` often indicates that the model may be overfitting, especially in cases where there’s no error. This means the model is capturing not only the underlying patterns in the data but also the noise, which may lead to poor generalization on unseen data.


In [8]:
print("Coefficients:", poly_model.coef_)
print("Intercept:", poly_model.intercept_)

equation = f"h(x) = {poly_model.intercept_}"
for i, coef in enumerate(poly_model.coef_[1:], 1): 
    equation += f" + {coef} * X^{i}"
print(equation)

Coefficients: [ 0.00000000e+00 -1.27196088e-07  1.20000000e+01  2.85714287e-02
  2.00000000e+00  1.26471056e-11]
Intercept: 2.0477862562984228e-05
h(x) = 2.0477862562984228e-05 + -1.2719608788458459e-07 * X^1 + 11.999999977692786 * X^2 + 0.028571428708642127 * X^3 + 2.000000000017223 * X^4 + 1.264710558501747e-11 * X^5


The equation for the polynomial model is:

$$
h(x) = 2.0478 \times 10^{-5} - 1.2719 \times 10^{-7} \cdot \text{KCl} + 12.0 \cdot \text{Temperature} + 0.0286 \cdot (\text{KCl})^2 + 2.0000 \cdot (\text{KCl} \cdot \text{Temperature}) + 1.2647 \times 10^{-11} \cdot (\text{Temperature})^2
$$

Where:
- The coefficients are approximated to five significant figures for clarity.
